fdbserver/DeltaTree.h

```diff
@@ -25,13 +25,27 @@
 #include "fdbclient/FDBTypes.h"
 #include "fdbserver/Knobs.h"
 #include <string.h>
+#include <emmintrin.h>
 
 typedef uint64_t Word;
 // Get the number of prefix bytes that are the same between a and b, up to their common length of cl
 static inline int commonPrefixLength(uint8_t const* ap, uint8_t const* bp, int cl) {
 	int i = 0;
+
+	for (; i + 16 < cl; i += 16) {
+		__m128i a = _mm_loadu_si128((__m128i*)(ap + i));
```

Comment:
Supposedly, `_mm_lddqu_si128` will be faster for misaligned strings (which strings should be in the common case, since things  higher in the tree would have had a common prefix).

Reply:
This sounds like the right choice.  While we could force one use case of this, RedwoodRecordRef, to use 16 byte aligned key strings when it allocates memory for reconstituted keys, that will waste some memory and we have other uses that likely won't be aligned.  For example, KeyValueStoreMemory uses this function to delta compress snapshots.

Reply:
To use that intrinsic or `_mm_cmpestri`, we would need to revert https://github.com/apple/foundationdb/pull/2414. @atn34

Reply:
> To use that intrinsic or `_mm_cmpestri`, we would need to revert #2414. @atn34

We could also move it to its own translation unit, and use different compilation flags just for that translation unit.

IIRC the original motivation for #2414 was as a potential mitigation for #2463, but we've since root caused #2463 and it was unrelated to alignment. FWIW I still think the reasoning described in comments in #2414 is sound

Reply:
AVX is enabled by default in master.  We could backport the change to 6.3 (with AVX disabled by default) if you want.

Reply:
Also, mixing "-m..." compilation flags between translation units isn't guaranteed to work.  They can change struct layouts and function calling conventions.  They usually don't, but mixing and matching is probably riskier than globally enabling everything. 

We used to require AVX by default, and AVX has been around for a very long time.  We can reasonably assume most intel machines support it.  The only place where it's tempting to mix and match is with the AVX512 extensions, but we're not using those for anything yet.

I have a branch somewhere that does runtime detection of CPU features, etc., but decided it wasn't worth the effort or risk to merge it.

Reply:
> Also, mixing "-m..." compilation flags between translation units isn't guaranteed to work. They can change struct layouts and function calling conventions.

Do you have a link I could read to learn more about this?